### PR TITLE
27 - Code refactoring to simplify integration with gui

### PR DESCRIPTION
**Pull request by balicz3k**
All algorithms are now encapsulated in dedicated classes that share common functionality through the BaseMethod superclass. Shared concerns (data loading, run pipeline, metrics, CSV output) are centralized, reducing duplication.

@danylo-denysov  FYI 
Each algorithm class defines its own PARAM_DEFS list, providing:

A clear, ordered specification of tunable parameters (name, type, defaults, bounds, description).
A uniform way to collect user input in the GUI.
Direct instantiation from an ordered values list without manual kwargs mapping.
The YAML file (descriptions.yaml) replaces the previous JSON file for method descriptions, allowing cleaner, readable multi‑line documentation and keeping source files lighter.

@kb-pk @danylo-denysov I would like to ask for a review, but potential comments can be solved by integrating with the GUI , so please specify if comments must be resolved before merging or may be solved in the next PR.

### CODE REVIEW

**Comment by kb-pk** [comment]
Thanks! Will look into that first thing in the morning

**Comment by balicz3k** [comment]
@danylo-denysov If you need some new additional functions. That may be helpful to integrate gui, please let me know.

**Comment by danylo-denysov** [comment]
I think we can move the plot method/methods to the base class, will be easier for implementing gui. Also the information that is being printed into console can be displayed in a "diagonstics" tab for example in the application.

**Comment by balicz3k** [comment]
ok, thx, I will try to move functions (useful from pov gui) from common to BaseMethodClass.

File: scheduler/Common.py [hunk 1]
```diff
@@ -141,3 +164,158 @@ def check_task_machine_mapping(machines, tasks, features, mappings):
...
+            
+            if output_mode == MAKESPAN_O_MODE:
+                print(f"    - Zadanie {task_id} (czas: {task_time:.2f})")
+            elif output_mode == ENERGY_O_MODE:
+                print(f"    - Zadanie {task_id} (energia: {task_energy:.2f})")
+            elif output_mode == ALL_O_MODE:
+                print(f"    - Zadanie {task_id} (czas: {task_time:.2f}, energia: {task_energy:.2f})")
+        print("-" * 20)
+
+def plot_gantt_chart(schedule_map, etc, makespan):
```
Comment (kb-pk) [inline scheduler/Common.py:265]: As ive said in the previous code review:

~~https://github.com/kb-pk/task-scheduling/pull/25 also implements this and i feel like this should be implemented on the GUI side rather than here.~~ @danylo-denysov has said that the functions can be implemented in the base class, so i guess we'll go with that

Also, makespan is not the only fitness metric that can be used to display a Gantt chart (e.g. total energy usage with seperate colors for active state and idle state), so again abstracting fitness function result into a class would be preferable

**Comment by balicz3k** [comment]
@kb-pk @danylo-denysov FYI, after integration with the GUI, I will implement the facade design pattern to avoid creating unnecessary objects .

File: scheduler/__main__.py [hunk 2]
```diff
@@ -2,47 +2,69 @@
...
 from .methods import Pitt_perm
 from .methods import Dragonfly
+from .methods import Fruitfly
+from .methods.BaseMethod import BaseMethod
 from . import Common
-from . import Utils
 
 import time
+from typing import Optional, Type
+from .Parametrs import instantiate_method
```
Comment (kb-pk) [inline scheduler/__main__.py:11]: These imports probably shouldn't be relative

File: scheduler/__main__.py [hunk 3]
```diff
@@ -2,47 +2,69 @@
...
+        run_algorithm(Michigan.MichiganMethod, values_list=[100, 0.01, False])
         time.sleep(1)
     elif x == 4:
-        Dragonfly.main(),
-        time.sleep(1)
+        run_algorithm(Dragonfly.DragonflyMethod, values_list=[100, 30, 0.9, 0.1, 0.1, 0.1, 2.0, 1.0, 0.5, False])
     elif x == 5:
-        Common.scheduling_mode = Common.ENERGY_MODE if Common.scheduling_mode == Common.MAKESPAN_MODE else Common.MAKESPAN_MODE
+        run_algorithm(Fruitfly.FruitflyMethod, values_list=[100, 10, 5, False])
+        time.sleep(1)
```
Comment (kb-pk) [inline scheduler/__main__.py:43]: Apart from being magic numbers, copy-pasting default values for parameters here for me is a bit pointless, since the job of default values are to be default and not needing to be specified

Comment (kb-pk) [reply scheduler/__main__.py:43]: Also, we should ideally let the CLI user tune the method's parameters instead of relying on the default ones

File: scheduler/methods/BaseMethod.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+class Lang(Enum):
+    PL = 0
+    EN = 1
+
+class BaseMethod(ABC):
+
+    _DESCRIPTIONS_CACHE = None
+
+    def __init__(self, iterations = 100, show_chart=True):
```
Comment (kb-pk) [inline scheduler/methods/BaseMethod.py:16]: Please make the style here consistent (`iterations=100`)

File: scheduler/methods/BaseMethod.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        """Konwersja solution -> {machine_id: [task_ids]}."""
+        ...
+
+    def run(self):
+        self.initialize()
+        self.optimize()
+        solution = self.get_best_solution()
+        schedule_map = self.build_schedule_map(solution)
+        makespan, total_energy = self._compute_metrics(schedule_map)
+        Common.print_schedule(schedule_map, self.etc, self.machines, makespan, total_energy)
```
Comment (kb-pk) [inline scheduler/methods/BaseMethod.py:80]: Make it just a `metrics` tuple, since we would have to change this code if we would ever need to add a new scheduling mode (e.g. security)

File: scheduler/Common.py [hunk 6]
```diff
@@ -141,3 +164,158 @@ def check_task_machine_mapping(machines, tasks, features, mappings):
...
+    :param etc_matrix: macierz ETC
+    :param machines_df: DataFrame maszyn
+    :param mode: MAKESPAN_MODE lub ENERGY_MODE
+    """
+    mk, en = compute_schedule_metrics(position_vector, etc_matrix, machines_df)
+    if mode == ENERGY_MODE:
+        return en, mk
+    return mk, en
+
+def print_schedule(schedule_map, etc, machines, makespan, total_energy):
```
Comment (kb-pk) [inline scheduler/Common.py:210]: As is said later, accept a tuple and not a named variable (`makespan`, `total_energy`), this will make adding a new scheduling method less painful

File: scheduler/Parametrs.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import Any, List, Optional, Type, Dict, Callable
+from .methods.BaseMethod import BaseMethod
```
Comment (kb-pk) [inline scheduler/Parametrs.py:4]: A relative import is probably not needed

File: scheduler/Common.py [hunk 8]
```diff
@@ -141,3 +164,158 @@ def check_task_machine_mapping(machines, tasks, features, mappings):
...
+    Zwraca (primary_metric, secondary_metric) zależnie od trybu (makespan / energy).
+    :param position_vector: wektor pozycji
+    :param etc_matrix: macierz ETC
+    :param machines_df: DataFrame maszyn
+    :param mode: MAKESPAN_MODE lub ENERGY_MODE
+    """
+    mk, en = compute_schedule_metrics(position_vector, etc_matrix, machines_df)
+    if mode == ENERGY_MODE:
+        return en, mk
+    return mk, en
```
Comment (kb-pk) [inline scheduler/Common.py:208]: A couple of things here:
- as per https://github.com/kb-pk/task-scheduling/issues/19, code returning a fitness function's value of an entity in a method should be implemented in that method's class (as is the case with `_compute_metrics()`!) - this code is common to all particle swarm methods (Dragonfly and Fruitfly), so a base class (e.g. `ParticleSwarmMethod`) would be preferable
- if you want to have all metrics returned by a single function, please make the functions calculating there metrics separate (e.g. `makespan()`, `energy()` or a basic `get_entity_fitness()` to use in a method)

File: scheduler/methods/Michigan.py [hunk 9]
```diff
@@ -1,516 +1,267 @@
 import csv
 from random import randint
-
 import numpy as np
 import pandas as pd
 from sklearn.utils import shuffle
-import scheduler.Utils as Utils
 
+from .BaseMethod import BaseMethod
+from .BaseMethod import Lang
```
Comment (kb-pk) [inline scheduler/methods/Michigan.py:8]: relative imports

File: scheduler/methods/Dragonfly.py [hunk 10]
```diff
@@ -1,201 +1,125 @@
...
+    def __init__(self,
+                 iterations=100,
+                 population_size=30,
+                 w_inertia=0.9,
+                 w_separation=0.1,
+                 w_alignment=0.1,
+                 w_cohesion=0.1,
+                 w_food=2.0,
+                 w_enemy=1.0,
+                 neighbour_radius_factor=0.5,
```
Comment (kb-pk) [inline scheduler/methods/Dragonfly.py:30]: _**This concerns all of the scheduling methods' classes**_, but you're repeating the **_default_** values three times (here, in the params and in the method class instantiation). Please fix this.

**Review by kb-pk — CHANGES_REQUESTED** [review]
Since this is a big PR, i have some requests. @balicz3k, please review and comment if necessary.

File: scheduler/methods/BaseMethod.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
+from abc import ABC, abstractmethod
+import scheduler.Common as Common
+import csv
+import numpy as np
+from enum import Enum
+import yaml, os
+
+class Lang(Enum):
+    PL = 0
+    EN = 1
```
[the remark this answers is not in the document]
Comment (balicz3k) [reply scheduler/methods/BaseMethod.py:10]: Enum used only with the Method classes I think it is a good place.

File: scheduler/Common.py (hunk 1 above)
Comment (balicz3k) [reply scheduler/Common.py:265]: I need your help with resolve this comment. Let me know when I can call you :)

File: scheduler/Common.py (hunk 6 above)
Comment (balicz3k) [reply scheduler/Common.py:210]: This method in the newest version do not need any parameters because now it is encapsulated into BaseMethod class

File: scheduler/methods/Dragonfly.py (hunk 10 above)
Comment (balicz3k) [reply scheduler/methods/Dragonfly.py:30]: As I said after introducing gui, default values should not be used anymore, so we can remove it later.

File: scheduler/methods/BaseMethod.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kb-pk) [reply scheduler/methods/BaseMethod.py:10]: Okay, for now it can stay, however when (if?) we add a translation layer to the CLI/GUI we'll need to move this

File: scheduler/Common.py [hunk 12]
```diff
@@ -141,3 +164,158 @@ def check_task_machine_mapping(machines, tasks, features, mappings):
...
 
+def decode_position_vector(position_vector, machines_num: int):
+    """
+    Zaokrąglenie i obcięcie wartości wektora pozycji do zakresu ID maszyn.
+    :param position_vector: ndarray float (długość = liczba zadań)
+    :param machines_num: liczba maszyn
+    :return: ndarray int (przypisanie task -> machine_id)
+    """
+    assign = np.rint(position_vector).astype(int)
+    return np.clip(assign, 0, machines_num - 1)
```
[the remark this answers is not in the document]
Comment (kb-pk) [reply scheduler/Common.py:175]: I still think that a base class with `decode()` would be better.
